### PR TITLE
Additional Stokes RHS

### PR DESCRIPTION
**Pull request by tjhei**
Latest idea on how to do force vector for the Stokes system.

supersedes #1300

### CODE REVIEW

**Comment by bangerth** [comment]
Can you update the test?

File: include/aspect/assembly.h [hunk 1]
```diff
@@ -711,6 +711,9 @@ namespace aspect
       public SimulatorAccess<dim>
     {
       public:
+
+        virtual void create_additional_material_model_outputs(MaterialModel::MaterialModelOutputs<dim> &outputs) const;
```
Comment (bangerth) [inline include/aspect/assembly.h]: what's this function gonna do?

File: include/aspect/material_model/interface.h [hunk 2]
```diff
@@ -516,6 +516,39 @@ namespace aspect
...
+        {
+          // TODO: not implemented
+        }
+
+        /**
+         * RHS (force) for the Stokes system with the following components:
+         * dim components for velocity, 1 component for pressure, 1 component
+         * for compaction pressure (if melt transport is enabled).
+         */
+        std::vector<std::vector<double> > rhs;
```
Comment (bangerth) [inline include/aspect/material_model/interface.h]: You call this class `StokesRHS`, so I suspect that the force terms should be `Tensor<1,dim>` objects, assuming you're talking about the force balance equation. If you want additional terms for the mass balance equation, maybe have another `std::vector<double>`.

If you really wanted to be generic and add terms for other equations as well, I would suggest to not document the order here, but use the same ordering as used by the introspection object (to which both the material models and assemblers have access).

File: source/simulator/assemblers/stokes.cc [hunk 3]
```diff
@@ -30,6 +31,22 @@ namespace aspect
     template <int dim>
     void
     StokesAssembler<dim>::
+    create_additional_material_model_outputs(MaterialModel::MaterialModelOutputs<dim> &outputs) const
+    {
+      if (this->get_parameters().enable_additional_stokes_rhs
+          && !outputs.template get_additional_output<MaterialModel::AdditionalMaterialOutputsStokesRHS<dim> >())
```
Comment (bangerth) [inline source/simulator/assemblers/stokes.cc]: I'll keep fighting against using `operator !` to compare pointers for non-Null-ness till I die! :-)

Please write the second condition as 
```
         && (outputs.template get_additional_output<MaterialModel::AdditionalMaterialOutputsStokesRHS<dim> >() != NULL)
```
or similar.

File: include/aspect/material_model/interface.h [hunk 4]
```diff
@@ -516,6 +516,39 @@ namespace aspect
...
+     * Stokes system, which can be attached to the
+     * MaterialModel::MaterialModelOutputs structure and filled in the
+     * MaterialModel::Interface::evaluate() function.
+     */
+    template<int dim>
+    class AdditionalMaterialOutputsStokesRHS: public AdditionalMaterialOutputs<dim>
+    {
+      public:
+        AdditionalMaterialOutputsStokesRHS(const unsigned int n_points,
+                                           const unsigned int n_stokes_components)
```
Comment (bangerth) [inline include/aspect/material_model/interface.h]: how could `n_stokes_components` ever be different than `dim+1`?

File: source/simulator/assemblers/stokes.cc [hunk 5]
```diff
@@ -113,6 +130,9 @@ namespace aspect
       const unsigned int stokes_dofs_per_cell = data.local_dof_indices.size();
       const unsigned int n_q_points    = scratch.finite_element_values.n_quadrature_points;
 
+      MaterialModel::AdditionalMaterialOutputsStokesRHS<dim>
+      *force = scratch.material_model_outputs.template get_additional_output<MaterialModel::AdditionalMaterialOutputsStokesRHS<dim> >();
```
Comment (bangerth) [inline source/simulator/assemblers/stokes.cc]: make it `const`.

File: source/simulator/assemblers/stokes.cc [hunk 6]
```diff
@@ -148,13 +168,29 @@ namespace aspect
...
+          // fill values from Stokes RHS additional outputs
+          Tensor<1,dim> force_u = numbers::signaling_nan<Tensor<1,dim> >();
+          if (force != NULL)
+            {
+              for (unsigned int d=0; d<dim; ++d)
+                force_u[d] = force->rhs[q][d];
+            }
+          const double force_p = (force!=NULL) ?
+                                 force->rhs[q][dim]
+                                 : std::numeric_limits<double>::quiet_NaN();
```
Comment (bangerth) [inline source/simulator/assemblers/stokes.cc]: why not a signaling NaN?

File: source/simulator/melt.cc [hunk 7]
```diff
@@ -88,6 +89,17 @@ namespace aspect
     MeltEquations<dim>::create_additional_material_model_outputs(MaterialModel::MaterialModelOutputs<dim> &outputs) const
     {
       MeltHandler<dim>::create_material_model_outputs(outputs);
+
+      if (this->get_parameters().enable_additional_stokes_rhs
+          && !outputs.template get_additional_output<MaterialModel::AdditionalMaterialOutputsStokesRHS<dim> >())
```
Comment (bangerth) [inline source/simulator/melt.cc]: same here

File: source/simulator/melt.cc [hunk 8]
```diff
@@ -88,6 +89,17 @@ namespace aspect
...
+        {
+          const unsigned int n_points = outputs.viscosities.size();
+          const unsigned int n_stokes_comp = dim + 2;
```
Comment (bangerth) [inline source/simulator/melt.cc]: why this many?

File: source/simulator/melt.cc [hunk 9]
```diff
@@ -215,6 +227,8 @@ namespace aspect
       const unsigned int p_c_component_index = introspection.variable("compaction pressure").first_component_index;
 
       MaterialModel::MeltOutputs<dim> *melt_outputs = scratch.material_model_outputs.template get_additional_output<MaterialModel::MeltOutputs<dim> >();
+      MaterialModel::AdditionalMaterialOutputsStokesRHS<dim>
+      *force = scratch.material_model_outputs.template get_additional_output<MaterialModel::AdditionalMaterialOutputsStokesRHS<dim> >();
```
Comment (bangerth) [inline source/simulator/melt.cc]: make `const`.

File: source/simulator/melt.cc [hunk 10]
```diff
@@ -293,13 +307,34 @@ namespace aspect
...
+            {
+              for (unsigned int d=0; d<dim; ++d)
+                force_u[d] = force->rhs[q][d];
+            }
+          const double force_p = (force!=NULL) ?
+                                 force->rhs[q][dim]
+                                 : std::numeric_limits<double>::quiet_NaN();
+          const double force_pc = (force!=NULL) ?
+                                  force->rhs[q][dim+1]
+                                  : std::numeric_limits<double>::quiet_NaN();
```
Comment (bangerth) [inline source/simulator/melt.cc]: signaling NaN?

File: source/simulator/parameters.cc [hunk 11]
```diff
@@ -546,6 +546,13 @@ namespace aspect
                          "\n\n"
                          "Note that while more than one operation can be selected it only makes sense to "
                          "pick one rotational and one translational operation.");
+      prm.declare_entry ("Enable additional Stokes RHS", "false",
+                         Patterns::Bool (),
+                         "Whether to ask the material model for additional terms for the right-hand side "
+                         "of the Stokes equation. This can be used to implement force vectors for "
+                         "manufactured solution problems. The assembler for the Stokes system will "
+                         "create an additional output of type AdditionalMaterialOutputsStokesRHS.");
```
Comment (bangerth) [inline source/simulator/parameters.cc]: the last sentence seems like a very detailed description of internals. there are many levels of knowledge to reach for a user to understand what this means.

File: include/aspect/material_model/interface.h (hunk 4 above)
Comment (tjhei) [inline include/aspect/material_model/interface.h]: Normally dim+1 but dim+2 if you have melt transport enabled. That is also the reason I did not want to use
```
Tensor<1,dim> u;
double p;
double p_c;
```
or something of that matter.

Comment (bangerth) [inline include/aspect/material_model/interface.h]: But then it's not just Stokes but everything.

**Comment by tjhei** [comment]
okay, ready for another round!

File: include/aspect/material_model/interface.h [hunk 12]
```diff
@@ -589,6 +589,53 @@ namespace aspect
...
+        {}
+
+        virtual ~AdditionalMaterialOutputsStokesRHS()
+        {}
+
+        virtual void average (const MaterialAveraging::AveragingOperation /*operation*/,
+                              const FullMatrix<double>  &/*projection_matrix*/,
+                              const FullMatrix<double>  &/*expansion_matrix*/)
+        {
+          // TODO: not implemented
```
Comment (tjhei) [inline include/aspect/material_model/interface.h]: need assert

**Comment by tjhei** [comment]
W writing: After further consideration, I'm merging from @tjhei's computer. The assertion cannot be added because the function is actually called.

**Pull request merged**